### PR TITLE
tasks: add tablet repair virtual task

### PR DESCRIPTION
**Pull request by Deexie**
Add tablet task manager module and keep it in storage_service.
Introduce tablet_virtual_task that covers tablet repair.

Thanks to a repair virtual task, a user can check the list of pending 
repairs, get the status of a specific repair, or abort it using the task 
manager API.

Fixes: #21368.

No backport, new feature

### CODE REVIEW

**Comment by Deexie** [comment]
- use coroutine::lambda in task_handler::abort

**Comment by Deexie** [comment]
- save captures in  task_handler::abort

File: service/task_manager_module.cc [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+future<std::optional<tasks::task_manager::virtual_task_hint>> tablet_virtual_task::contains(tasks::task_id task_id) const {
+    auto tables = get_table_ids();
+    for (auto table : tables) {
+        bool has_task = false;
+        auto& tmap = _ss.get_token_metadata().tablets().get_tablet_map(table);
+        co_await tmap.for_each_tablet([&] (locator::tablet_id tid, const locator::tablet_info& info) {
+            has_task = has_task || (info.repair_task_info.is_valid() && info.repair_task_info.tablet_task_id.uuid() == task_id.uuid());
+            return make_ready_future();
```
Comment (asias) [inline service/task_manager_module.cc:28]: We can stop the loop earlier if we found the task id. It is wasteful to loop all the tablets. For each tablet is not a good fit here.

File: tasks/task_manager.hh [hunk 2]
```diff
@@ -70,6 +70,7 @@ public:
     using foreign_task_map = std::unordered_map<task_id, foreign_task_ptr>;
     using module_ptr = shared_ptr<module>;
     using modules = std::unordered_map<std::string, module_ptr>;
+    using virtual_task_hint = std::unordered_map<sstring, sstring>;
```
Comment (asias) [inline tasks/task_manager.hh:73]: We know exactly what we want to put into the virtual_task_hint. Isn't it better to use the real type

struct virtual_task_hint {
std::optional<table_id> table_id;
std::optional<other_thing_for_the_hint> foo;
}

File: test/topology_tasks/test_tablet_tasks.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    token = -1
+    async def repair_task():
+        # Keep retring tablet repair.
+        await inject_error_on(manager, "repair_tablet_fail_on_rpc_call", servers)
+        await manager.api.tablet_repair(servers[0].ip_addr, "test", "test", token)
+
+    await asyncio.gather(repair_task(), check_and_abort_repair_task(tm, servers, module_name))
+
+    await cql.run_async("DROP KEYSPACE test;")
```
Comment (asias) [inline test/topology_tasks/test_tablet_tasks.py:73]: We can drop the drop it is useless for the test.

File: test/topology_tasks/test_tablet_tasks.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    status = await tm.get_task_status(servers[0].ip_addr, task.task_id)
+
+    check_repair_task_status(status)
+    assert status.state in ["created", "running"]
+
+    async def wait_for_task():
+        status_wait = await tm.wait_for_task(servers[0].ip_addr, task.task_id)
+        check_repair_task_status(status_wait)
+        assert status_wait.state == "done"
```
Comment (asias) [inline test/topology_tasks/test_tablet_tasks.py:49]: So the task manager could not know the task is aborted yet.  It could be a follow up.

File: test/topology_tasks/suite.yaml [hunk 5]
```diff
@@ -6,6 +6,7 @@ extra_scylla_config_options:
     authenticator: AllowAllAuthenticator
     authorizer: AllowAllAuthorizer
     enable_user_defined_functions: False
-    experimental_features: ['consistent-topology-changes', 'tablets']
+    experimental_features: ['consistent-topology-changes']
+    enable_tablets: True
```
Comment (asias) [inline test/topology_tasks/suite.yaml:10]: Better to split into a separate patch and explain why.

File: node_ops/task_manager_module.cc [hunk 6]
```diff
@@ -81,7 +81,7 @@ future<std::optional<tasks::task_status>> node_ops_virtual_task::get_status_help
         .kind = tasks::task_kind::cluster,
         .scope = "cluster",
         .state = get_state(entry),
-        .is_abortable = co_await is_abortable(),
+        .is_abortable = co_await is_abortable({}),
```
Comment (asias) [inline node_ops/task_manager_module.cc:84]: It is hard to know what the {} is for. We can make is_abortable() get a default empty task hint if nothing is passed.

File: service/task_manager_module.cc [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    }
+    co_return std::nullopt;
+}
+
+future<tasks::is_abortable> tablet_virtual_task::is_abortable(tasks::task_manager::virtual_task_hint hint) const {
+    return make_ready_future<tasks::is_abortable>(tasks::is_abortable::yes);
+}
+
+future<std::optional<tasks::task_status>> tablet_virtual_task::get_status(tasks::task_id id, tasks::task_manager::virtual_task_hint hint) {
+    auto table = table_id{utils::UUID{hint.at(TABLE_ID)}};
```
Comment (asias) [inline service/task_manager_module.cc:43]: What guarantees TABLE_ID is in the hint? (Perhaps the caller always provide one, but still it is better to check)

If we have

struct virtual_task_hint {
std::optional<table_id> table_id;
std::optional<other_thing_for_the_hint> foo;
}

we could do

if (hint.table_id) {
}

File: service/task_manager_module.cc [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    co_return status;
+}
+
+future<> tablet_virtual_task::abort(tasks::task_id id, tasks::task_manager::virtual_task_hint hint) noexcept {
+    auto table = table_id{utils::UUID{hint.at(TABLE_ID)}};
+    co_await _ss.del_repair_tablet_request(table, locator::tablet_task_id{id.uuid()});
+}
+
+future<std::vector<tasks::task_stats>> tablet_virtual_task::get_stats() {
+    std::vector<tasks::task_stats> res;
```
Comment (asias) [inline service/task_manager_module.cc:75]: Can we add test for the get_stats as well?

File: service/task_manager_module.cc [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+            auto task_id = tasks::task_id{info.repair_task_info.tablet_task_id.uuid()};
+            tasks::task_stats stats{
+                .task_id = task_id,
+                .type = locator::tablet_task_type_to_string(info.repair_task_info.request_type),
+                .kind = tasks::task_kind::cluster,
+                .scope = info.repair_task_info.is_user_request() ? "table" : "tablet",
+                .state = tasks::task_manager::task_state::running,
+                .keyspace = schema->ks_name(),
+                .table = schema->cf_name()
+            };
```
Comment (asias) [inline service/task_manager_module.cc:96]: We do not have the task create time / elapsed time in the task_stats? Those sound useful.

File: tasks/task_manager.hh (hunk 2 above)
Comment (Deexie) [reply tasks/task_manager.hh:73]: We do know what we want to use for repair tasks but `virtual_task_hint` can be used for all virtual tasks. We could alternatively have hint inheritance and cast the type for each task type, but I like the map more.

**Comment by Deexie** [comment]
- pass a hint to node_ops_virtual_task::is_abortable
- add virtual_task::get_hint_value
- return when the task is found in tablet_virtual_task::contains
- test:
	- enable tablets in topology_tasks tests in a separate commit
	- drop keyspace drop
	- add test for repair task stats

File: service/task_manager_module.cc (hunk 7 above)
Comment (Deexie) [reply service/task_manager_module.cc]: I changed it to `on_internal_error` if TABLE_ID isn't there. This situation should not occur

File: service/task_manager_module.cc (hunk 9 above)
Comment (Deexie) [reply service/task_manager_module.cc:92]: I agree. I'll change that separately.

**Comment by Deexie** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13304/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13304/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_tasks/test_tablet_tasks ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/322/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12923/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1818/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4918/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13304/)
> #### Build Details:
> 
>     * Duration: 8 hr 2 min
> 
>     * Builder: i-09a703c76cfdf938d (m5ad.8xlarge)

https://github.com/scylladb/scylladb/issues/21490

**Comment by avikivity** [comment]
> Add tablet task manager module and keep it in storage_service. Introduce tablet_virtual_task that covers tablet repair.
> 


Please add an explanation about why we're doing this (and what is improved from the user's point of view)

File: tasks/task_manager.hh (hunk 2 above)
Comment (asias) [reply tasks/task_manager.hh:73]: The problem with the string map is that we have to serialize and deserialize to and from string unnecessarily. 

We can use any tasks we want with the following, without even playing with inheritance and type casks. We can simply add only types that we need for different tasks. Some of fields can be used by different tasks.
```
struct virtual_task_hint {
  std::optional<table_id> table_id;
  std::optional<foo> foo;
  ...
  std::optional<bar> bar;
}

if (!virtual_task_hint.foo) {
     // throw error if foo is expected
}
// use foo after the check
```

**Comment by Deexie** [comment]
- virtual_task_hint is a struct with optionals
	- drop task group formatter
- pass hint to `tablet_virtual_task::get_status_helper`
- fix test (it is acceptable that one node shows that the repair state is "created" and the other - "running")
- rebase
- update cover letter

File: service/task_manager_module.cc [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        res.state = sched_nr == 0 ? tasks::task_manager::task_state::created : tasks::task_manager::task_state::running;
+        co_return res;
+    }
+    // FIXME: Show finished tasks.
+    co_return std::nullopt;
+}
+
+table_id tablet_virtual_task::get_table_id(const tasks::virtual_task_hint& hint) const {
+    if (!hint.table_id.has_value()) {
+        on_internal_error(tasks::tmlogger, "tablet_virtual_task hint does not contain table_id");
```
Comment (denesb) [inline service/task_manager_module.cc:151]: If this `table_id` is expected to be always present, then why make it optional?

File: service/task_manager_module.cc [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+                co_return tasks::virtual_task_hint{
+                    .table_id = table,
+                };
+            }
+        }
+    }
+    co_return std::nullopt;
+}
+
+future<tasks::is_abortable> tablet_virtual_task::is_abortable(tasks::virtual_task_hint hint) const {
```
Comment (denesb) [inline service/task_manager_module.cc:37]: `hint` param is unused, will it be used in a later patch?

File: service/task_manager_module.cc [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+            if (!info.repair_task_info.is_valid()) {
+                return make_ready_future();
+            }
+
+            auto task_id = tasks::task_id{info.repair_task_info.tablet_task_id.uuid()};
+            tasks::task_stats stats{
+                .task_id = task_id,
+                .type = locator::tablet_task_type_to_string(info.repair_task_info.request_type),
+                .kind = tasks::task_kind::cluster,
+                .scope = info.repair_task_info.is_user_request() ? "table" : "tablet",
```
Comment (denesb) [inline service/task_manager_module.cc:90]: User request always means entire table is repaired? I thought we allow for repairing a single tablet too.

File: service/task_manager_module.cc (hunk 10 above)
Comment (Deexie) [reply service/task_manager_module.cc:151]: It's always present for tablet virtual tasks, but virtual_task_hint is meant for all virtual tasks

File: service/task_manager_module.cc (hunk 11 above)
Comment (Deexie) [reply service/task_manager_module.cc:37]: In this PR - no. It will be used later though. I can remove it for now, but I think it makes sense to pass the hint to all public virtual task methods

File: service/task_manager_module.cc (hunk 12 above)
Comment (Deexie) [reply service/task_manager_module.cc:90]: We can repair a single tablet. User request specifies a list of tokens to repair. It is not always the whole table, but it is at most one table, and hence the scope

File: service/task_manager_module.cc (hunk 11 above)
Comment (denesb) [reply service/task_manager_module.cc:37]: Ok, it can stay if you have plans for it.
When this is the case, best to mention this in the commitlog, so reviewers are not left guessing if this is a mistake or part of a plan.

File: service/task_manager_module.cc (hunk 12 above)
Comment (denesb) [reply service/task_manager_module.cc:90]: Ok.

File: service/task_manager_module.cc (hunk 10 above)
Comment (denesb) [reply service/task_manager_module.cc:151]: Ok.

File: service/task_manager_module.cc [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+tasks::task_manager::task_group tablet_virtual_task::get_group() const noexcept {
+    return tasks::task_manager::task_group::tablets_group;
+}
+
+future<std::optional<tasks::virtual_task_hint>> tablet_virtual_task::contains(tasks::task_id task_id) const {
+    auto tables = get_table_ids();
+    for (auto table : tables) {
+        auto& tmap = _ss.get_token_metadata().tablets().get_tablet_map(table);
+        for (const locator::tablet_info& info : tmap.tablets()) {
+            if (info.repair_task_info.is_valid() && info.repair_task_info.tablet_task_id.uuid() == task_id.uuid()) {
```
Comment (asias) [inline service/task_manager_module.cc:27]: We need to yield, otherwise we might stall here, e.g., use coroutine::maybe_yield().

File: test/topology_tasks/test_tablet_tasks.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        assert task0.table in ["test", "test2", "test3"]
+        assert task0.table == task1.table, f"Inconsistent table for task {task0.task_id}"
+        assert task0.table == task2.table, f"Inconsistent table for task {task0.task_id}"
+
+        for task in [task0, task1, task2]:
+            assert task.state in ["created", "running"]
+            assert task.type == "user_repair"
+            assert task.kind == "cluster"
+            assert task.scope == "table"
+            assert task.keyspace == "test"
```
Comment (asias) [inline test/topology_tasks/test_tablet_tasks.py:104]: nit: Perhaps we can reuse check_repair_task_status above, e.g., 

check_repair_task_status(expected_states(['created', 'running']))

File: test/topology_tasks/test_tablet_tasks.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+async def check_repair_task_list(tm: TaskManagerClient, servers: list[ServerInfo], module_name: str):
+    def get_task_with_id(repair_tasks, task_id):
+        tasks_with_id1 = [task for task in repair_tasks if task.task_id == task_id]
+        assert len(tasks_with_id1) == 1
+        return tasks_with_id1[0]
+
+    # Wait until user repair tasks are created.
+    repair_tasks0 = await wait_for_tasks(tm, servers[0], module_name, len(servers))
+    repair_tasks1 = await wait_for_tasks(tm, servers[1], module_name, len(servers))
+    repair_tasks2 = await wait_for_tasks(tm, servers[2], module_name, len(servers))
```
Comment (asias) [inline test/topology_tasks/test_tablet_tasks.py:86]: I just want to confirm that all the nodes are supposed to get same repair tasks, no matter where the repair task is created and where the repair task is queried.

File: tasks/task_manager.hh [hunk 16]
```diff
@@ -272,14 +274,14 @@ public:
             static future<std::vector<task_identity>> get_children(module_ptr module, task_id parent_id);
         public:
             virtual task_group get_group() const noexcept = 0;
-            virtual future<bool> contains(tasks::task_id task_id) const = 0;
+            virtual future<std::optional<virtual_task_hint>> contains(tasks::task_id task_id) const = 0;
```
Comment (asias) [inline tasks/task_manager.hh:278]: Lets' add a doc for the this function for the difference between empty virtual_task_hint and std::nullopt.

**Comment by asias** [comment]
Looks good to me. I have few minor comments only.

File: test/topology_tasks/test_tablet_tasks.py (hunk 14 above)
Comment (Deexie) [reply test/topology_tasks/test_tablet_tasks.py:104]: Those are different types (TaskStatus vs TaskStats) and I would rather not mix them

**Comment by Deexie** [comment]
- add maybe_yield to tablet_virtual_task::contains
- add doc about virtual_task::impl::contains
- test:
	- rename `wait_for_task` to `wait_tasks_created`
	- add states param to `check_repair_task_status`

File: test/topology_tasks/test_tablet_tasks.py (hunk 14 above)
Comment (asias) [reply test/topology_tasks/test_tablet_tasks.py:104]: ok

File: test/topology_tasks/test_tablet_tasks.py (hunk 15 above)
Comment (asias) [reply test/topology_tasks/test_tablet_tasks.py]: Ok

**Comment by bhalevy** [comment]
@denesb / @scylladb/scylla-maint please consider merging
